### ISSUE TITLE
Fiberassign to recognize SV mask

### ISSUE THREAD

**Issue by Srheft** [desihub/fiberassign#181]
@tskisner @sbailey:

It appears that the current Fiberassign code only recongnizes the main survey masks:
[https://github.com/desihub/fiberassign/blob/3beff2b8d484233fd705b33021e7f6972f11dcc2/py/fiberassign/assign.py#L29](url) 

If there is already a way to exchange the masks via "--options", please let me know and if the comment is valid (test materials are [here](http://www.astro.utah.edu/~u6022465/test/)), please accomodate the means to run fiberassign for SV targets (includes having `from desitarget.sv1.sv1_targetmask import desi_mask` ).

In my experience, [running fiberassign](http://www.astro.utah.edu/~u6022465/test/FA.log) with this MTL file ([MTL_DEEP2_1.fits](http://www.astro.utah.edu/~u6022465/test/)) from the SV catalog returns an ALL SKY assignment while the field is quite crowded with science targets of equaly high priority:

![image](https://user-images.githubusercontent.com/32471902/52898206-3b3aa580-3198-11e9-9639-1e81bd944ef8.png)

If you have a working example that proves otherwise, please share it here as I did not find any issue with the input files I used. Thanks.

**Comment by sbailey** [comment]
fiberassign does not currently support this (but it should).  I think there are two pieces to this:
* what column name should be used for the targeting mask bits (default DESI_TARGET)?
* what should the default masks be for SCIENCE, STDSTAR, SKY, SAFE, and exclude locations, since the exactly bits will be different for different programs?

In the code we could add a --program option (default DESI, but also SV1, CMX, ...) and pass that as arguments to `fiberassign.targets.get_sciencemask()`, `.get_stdmask()` etc.

Opinions welcome if this really is "program" or if a different word should be used.

@Srheft could you please fill out / check these details:

Defaults for program DESI:
  * column name: `DESI_TARGET`
  * mask name: `desitarget.targetmask.desi_mask`
  * SCIENCE mask: `ELG|LRG|QSO|BGS_ANY|MWS_ANY|SECONDARY_ANY`
  * STD mask: `STD_FAINT|STD_BRIGHT|STD_WD`
  * SKY mask: `SKY`
  * SAFE mask: `BAD_SKY`
  * EXCLUDE mask: `BRIGHT_OBJECT|IN_BRIGHT_OBJECT`

Defaults for program SV1:
  * column name: `SV1_TARGET`
  * mask name: `desitarget.sv1_targetmask.desi_mask`
  * SCIENCE mask: `ELG|LRG|QSO|BGS_ANY|MWS_ANY|SECONDARY_ANY`
  * STD mask: `STD_FAINT|STD_BRIGHT|STD_WD`
  * SKY mask: `SKY`
  * SAFE mask: `BAD_SKY`
  * EXCLUDE mask: `BRIGHT_OBJECT|IN_BRIGHT_OBJECT`

Defaults for program CMX:
  * column name: `CMX_TARGET`
  * mask name: `desitarget.cmx_targetmask.cmx_mask`
  * SCIENCE mask: ???
  * STD mask: ???
  * SKY mask: ???
  * SAFE mask: ???
  * EXCLUDE mask: ???

Other programs?

Also mentioning @geordie666 since some of these concepts may not yet be fleshed out for CMX...

Detail: it would be handy if we could use SV1_TARGET for an SV science target list while re-using DESI_TARGET for a SKY file; I'm not sure the best way to support that so throwing it out there for discussion.

**Comment by Srheft** [comment]
@sbailey:  I have not had any issue with running Fiberassign on CMX catalog [hence the plots today] and I couldn't tell why the same sterategy for CMX_TARGE, did not work with SV1_DESI_TARGET (in your note, SV1_TARGET  --> SV1_DESI_TARGET - was your list informed by [this](https://github.com/desihub/desitarget/blob/master/py/desitarget/sv1/data/sv1_targetmask.yaml)?).

Looking at  [https://github.com/desihub/desitarget/blob/master/py/desitarget/cmx/data/cmx_targetmask.yaml](url)

Defaults for program CMX (@geordie666: please verify):

* column name: `CMX_TARGET`
* mask name: `desitarget.cmx_targetmask.cmx_mask`
* SCIENCE mask: `STD_GAIA|SV0_STD_BRIGHT|STD_TEST|STD_CALSPEC|SV0_BGS|SV0_MWS`
* STD mask: NONE specified --> perhaps means borrowing from main survey mask: `STD_FAINT|STD_BRIGHT|STD_WD`
* SKY mask: NONE specified --> perhaps means borrowing from the main survey mask: `SKY`
* SAFE mask:NONE specified --> perhaps means borrowing from the main survey SAFE mask: `BAD_SKY`
* EXCLUDE mask:  NONE specified

I was hoping that the mask exchange could be triggered in the --option for fiberassign. 
As for  your question on "other programs?", I don't think so but @kdawson1000 would know best  [I take it that the initial SV selection (SV0) does not count as an observing  "program"].

**Comment by tskisner** [comment]
This is a more general question:  Do we want the `fba_run` executable to know anything about these programs?  Would it make more sense to have very short, specialized scripts with a limited set of options which call the high level fiberassign methods directly?  For example, if we are assigning only sky fibers during some test, we don't need to go through all the no-op code of trying to assign science targets when none exist.  I fear that we will quickly explode the commandline options with specialized flags for a specific program, and then carry all that stuff around inside a single script when we don't need to.

**Comment by Srheft** [comment]
@tskisner: 

- Agreed that too many flags is not ideal [used to call flags "band-aid" in quicklook]. It seemed to be less invasive to the algorithm when a flag activates a mask exchange. Any solution you prefer is perfectly fine. Our objective is to give 5 initial runs on five SV fields for first round of analysis to SV team. 

- The nature of SV assignment is quite similar to the main survey assignment. All the functionalities of fba_run would be used in all SV assignments.

- ALL-SKY assignment is one of the CMX assignments that is already verified to be done with current fiberassign code. This ticket is just to make fba_run to recognize SV targets. I think CMX is suggested to be added to the list of "other programs" for completeness. 

@sbailey: 

> In the code we could add a --program option (default DESI, but also SV1, CMX, ...) and pass that as arguments to `fiberassign.targets.get_sciencemask()`, `.get_stdmask()` etc.
Opinions welcome if this really is "program" or if a different word should be used.

If the vote was to go with flags: the key 'PROGRAM' has been used for distinguishing DARK/GRAY/BRIGHT conditions in raw fits headers so if we don't want to touch that established convention, how about "--survey" or "--obsprogram" for distinguishing SV, CMX and DESI[default]?

**Comment by geordie666** [comment]
Not that we have to be consistent between fiber assignment and target selection, but I've been using "--survey" and SV, CMX and MAIN (instead of DESI).

**Comment by Srheft** [comment]
Good to know. Thanks. Interestingly I actually first wrote SV, CMX, MAIN but then changed it to DESI.

**Comment by sbailey** [comment]
*If* we go with flags (for fba_run or a different script), good for --survey with values SV, CMX, MAIN.

I like @tskisner's suggestion of keeping fba_run reasonably simple/straightforward, and  taking special cases to separate scripts / notebooks.

**For commissioning** fiber assignment is so different from the main survey that completely different custom scripts/notebooks makes sense.  

**For SV**, fiber assignment is very similar to the main survey so let's *not* have a `sv_fba_run` that is nearly identical to `fba_run` and differs only by what mask is loaded and the sky and stdstar default densities.  In that case it would be better to have `sv_fba_run` call `fba_run` with non-default options, so that any changes to `fba_run` itself are effectively also included in `sv_fba_run`.

To support different masks, I think the following would work:
* Add an option to fba_run for what column should be used for the target mask (default `DESI_TARGET`) and propagate that to `fiberassign.targets.load_target_file(..., typecol=...)`.
* wrapper scripts like `sv1_fba_run` could then call `fba_run` with non-default --typecol, and override --sciencemask, --stdmask, --skymask, --safemask, and --exclusemask with *numeric* values for the masks approproate for that typecol.

i.e. the SV-specific logic is isolated to an SV-specific script, with only a small update needed to fba_run itself to add a `--typecol` option (or similar name).

Thoughts?

**Comment by tskisner** [comment]
Is it reasonable to take the "guts" of fba_run and make it into a function (say ```fiberassign.assign.run()``` or similar)?  Actually that file is pretty long, so we could put it somewhere else...  Then we could call that high level function from both fba_run and other customized front-ends.

**Comment by tskisner** [comment]
I believe this issue was resolved by PR #188.  In particular, one can now use the "high level" interface (see the Example):

https://fiberassign.readthedocs.io/en/latest/reference.html#high-level-interface

To pass options to a function which runs the assignment.  This is convenient for running from a notebook or other script.  One can now fully override the column used for masking (`--mask_column`) as well as all the individual bit masks for the different types of targets.  Hopefully this is sufficient.  @Srheft , in another ticket you had a failing case of using the `--mask_column` option to specify a different column name.  If that still fails with latest master, then we should fix that before closing this issue.

**Comment by tskisner** [comment]
I tested the dataset from @Srheft and reproduced the error with `--mask_column` set to SV1_DESI_TARGET.  The problem is due to the fact that the specified column name is used for all target files.  It worked without error on the MTL_DEEP2_1.fits file, but the sky target file uses the DESI_TARGET column.

So, we either need to specify the column separately for each target file or force all target files used for a specific program to use the same column name.

Note that once https://github.com/desihub/desitarget/issues/464 is implemented, we can add a `--program` option to fba_run to automatically choose the default bitmasks for all target files.  However, that requires all target files to use the same column names and bit definitions.

**Comment by Srheft** [comment]
Thank you for verification @tskisner. I take it that the immediate solution [to run master on SV right now] is to specify --mask_column for each target file. Did you verify that

`--targets ./MTL_DEEP2_1.fits  --mask_column SV1_DESI_TARGET ./skies-dr7.1-PR459.fits --mask_column DESI_TARGET ` 

could work?  the code runs [with the syntax change above] but only skies are assigned (see [qa.json](http://www.astro.utah.edu/~u6022465/SV/18pRosette_SV_DEEP2_1/qa.json)).

**Comment by tskisner** [comment]
No, I'm saying that currently the only workaround is to have all target files use a single column name.  There is no way to specify a different column name / program per file.  In the future, we were thinking of having something like:
```
fba_run --progam SV1 --targets sv1.fits sky.fits ...
```
But that would not work unless all target files have the same SV1_DESI_TARGET column.  If we want to use target files (like sky) for multiple programs, then we will need to specify the program / column name individually per file.

Perhaps something like:
```
fba_run --targets sky.fits:main sv1_targets.fits:sv1 sv1_standards.fits:sv1
```
Or similar.

**Comment by geordie666** [comment]
For what it's worth, I don't think we'll ever have a "special" skies file for SV. So, the skies file could just be hard-coded to always have the "DESI_TARGET" column.  I don't see when we'd experiment with a substantially different set-up for skies for SV, and, if we did, git versioning should take care of how the algorithm was different.

In fact, I should probably deprecate the sky-related bits in the SV1 yaml file and use those bits, if needed, for other target classes.

**Comment by Srheft** [comment]
Naive question here: What if sky was not passed as --targets but rather as --skyfile like the old version of fiberassign? because SKY and BAD_SKY [safemask] are assigned at last, would the algorithm change much if we don't pass skies as --targets [completely ignore them and their bits until after we are done with --targets; then introduce --skyfile as *new targets* and their bits].
Basically running the same algorithm twice within one command & in this order: once on science targets and then on sky. I wonder if this could work better than the --program idea.

**Comment by kdawson1000** [comment]
My understanding from Ted is this:  use the target selection for skies in the same manner for commissioning, SV, and the main survey.  For that selection of sky targets, simply duplicate the DESI_TARGET column to also include a SV1_DESI_TARGET and a CMX_TARGET column, with bits within each bitmask assigned according to a sky designation.  

If I understand this correctly, this will be a trivial fix and will allow us to run fba using only a single mask_column based on what the science targets need.

**Comment by tskisner** [comment]
All of these issues can be fixed very easily- this is just a matter of how we parse arguments.  We just need to decide what we want.  It sounds like we need to specify / detect for **every target file individually**:

1.  The column name to use.

2.  The bitmasks for science, standards, sky, "safe", and the "exclusion mask".

The bitmasks in number 2 should be set to defaults provided by desi_target for a given program.  I suggest that in desi_target for each of these programs we define not only the default bitmasks for every program but also the column name used for that program.

Then in fiberassign, for each target file we can loop over the column names for each defined program to identify the program for that file and then what the default bitmasks should be.

Even if fiberassign can use functions in desi_target to identify the program for each target file and its defaullt bitmasks, we may still want to allow the user to override the column name and individual bitmasks for each type of target.  This kind of override gets too complicated for command line options, so we might have to introduce an optional config file format for complicated configurations like that.

Let's pause this ticket until we get the helper functions / dictionary in desi_target to query information about specific programs.  Then we can quickly set up the default behavior in fiberassign that works for 95% of the cases.

**Comment by geordie666** [comment]
@tskisner: The `main_cmx_or_sv(targets)` function in desitarget.targets:

https://github.com/desihub/desitarget/blob/master/py/desitarget/targets.py#L286

Will detect the column name and load the correct masks.

**Comment by tskisner** [comment]
Awesome!  That is exactly what we need.  Here is what I propose:

1.  By default, every target file should have column names and bitmasks that match one of the programs defined in desi_target.  So for example a sky target file or science target file would have column DESI_TARGET and the correct bits set.  An SV target file would have a column SV1_DESI_TARGET and the correct bits set.  These files can be mixed in the same run, since the information is obtained from desi_target for each file individually.

2.  For extreme manual intervention (i.e. you are doing something completely different and not using any programs defined in desi_target), then you must use the same column name and bit definitions across all target files, and must specify the `--mask_column` and the options for the 4 bit masks.

Does that meet our requirements?

**Comment by tskisner** [comment]
Hello @geordie666 , I looked into using your function in fiberassign, and I think it is an important part of the solution.  The thing we need in fiberassign for a given target file are:

1.  What column to use

2.  What integer masks to bitwise-and against this column to determine whether to treat an object as a "science" target, a "sky" target, a "standard" target, or a "safe" target.

Your function gives us 1 and part of 2 (the named bits in the column).  The final piece we need is the default selection of which bits to use for each of the 4 categories above.  For example, the current code sets the default integer mask for considering things as "science" targets:

https://github.com/desihub/fiberassign/blob/master/py/fiberassign/targets.py#L40

And similarly for the other categories.  It seems like we need that same functionality for the "sv" and "cmx" file types.

Questions for everyone:

1.  Looking at the yaml files, the SV bit names (at least the names we care about for assignment) seem to be the same as for the main survey.  Is this true?

2.  Looking at the cmx yaml file, my guess is that all of these object types should be handled as "science", since during commissioning you really just want to assign the targets you specify and then fill everything else with sky.  So there are no "fiberassign standards" during commissioning.  Correct?

If you agree with these last 2 points, then I'll use the same mapping of bitnames --> target categories for SV and main, and then treat **all** named cmx bits as "science".

**Comment by geordie666** [comment]
@tskisner: I believe you are correct on your first point, in that the critical names and bits`LRG`, `ELG`, `QSO`, `LRG_1PASS`, `LRG_2PASS` and the `STD` bits are the same for SV and the main survey. If we did update those bits for SV, I suspect that we'd also have to update them for the main survey.

One question for you: Can I deprecate the `SKY` and `BAD_SKY` bits in the SV yaml file? I think we'll only ever be reading those bits from the main survey mask, as we won't have a special format for SV sky locations (we'll let git versioning handle any updates to sky selection moving from SV to the main survey).

On your second point, I *think* that's correct, but I'm out of the loop as to whether any commissioning tests will require treating targets in a special way beyond classifying them as science targets. Perhaps @crockosi could comment on that?

**Comment by tskisner** [comment]
@geordie666 , I think that all the SV target files I have seen so far include only science targets (not sky) so I don't think there is any impact.  However, not sure about BAD_SKY.  In the past those targets were sometimes placed in the science target files and sometimes in the sky target files.

I'll open a PR shortly with changes that should resolve this issue for the current state of the file examples I have seen "in the wild".

**Comment by kdawson1000** [comment]
@geordie666 , @tskisner ,
I am trying to piece together the plan here.  I think I prefer to move the logic that maps bitnames --> target categories into an external yaml file, similar to what Adam has.  This will make it easier to deal with changes in the target bits and definitions, which are inevitable.  It will also facilitate the situation where the bits in the SV mask do not align with those in the main survey, although I do think we can keep those in line.

For target bits, we are talking about extending the SV selections to allow sub-categories of each target class.  For example, there will be red ELGs, faint ELGs, blue ELGs...I don't think that will interfere with the logic that Ted/Adam are proposing as long as all sub-categories of targets share the main ELG, LRG, or QSO flag.  In other words, the sub-category would be identified as an additional classification.

For CMX, it seems that we should still want std stars, so we should be asking for science, std star, and sky targets.

-Kyle

**Comment by sbailey** [comment]
I think everything here was addressed in PR #193.  Summary: --targets can accept files with DESI_TARGET, SV1_DESI_TARGET, or CMX_TARGET, as long as all files use the same column name.  It uses `desitarget.targets.main_cmx_or_sv` to know which to use.  The --skies option allows you to pass in a normal skies file with DESI_TARGET, even if your targets files are from SV1_DESI_TARGET or CMX_TARGET.

Any objections to closing this ticket?
